### PR TITLE
Added preferred_element_type arg to nnx.Linear*, nnx.Conv*, nnx.Einsum

### LINKED ISSUE #4890: Add `preferred_element_type` to `nnx.Linear` and other relevant layers
### 🚀 Feature request

A common mixed-precision pattern is to perform *heavy* computations (e.g., matrix multiplications in `Linear` layers) in `jnp.bfloat16` or `jnp.float16`, but then accumulate the end result in `jnp.float32` for subsequent operations, such as activation functions or normalization layers, to preserve precision.

Right now, achieving this behavior in Flax requires overriding `dot_general` or `conv_general_dilated` to set the `preferred_element_type`, which is cumbersome. Most of the relevant keyword arguments (e.g., `dtype`, `precision`) are already exposed in `__init__`, so the only missing piece is **`preferred_element_type`**.

---

### 💡 Proposed solution

Expose a `preferred_element_type` argument in layers like `nnx.Linear` and pass it through to the underlying JAX calls (`lax.dot_general`, `lax.conv_general_dilated`, etc.).  This would make mixed-precision setups significantly easier and avoid custom overrides.

---

### 📋 Minimal example

The example below demonstrates why `preferred_element_type` is necessary for performing computations in the intented precision with the expected runtime improvements. There might be a different way to achive the same, but I do not see it.

```python
import jax
import jax.numpy as jnp
from flax import nnx
from functools import partial

linear = jax.jit(nnx.Linear(5096, 5096, rngs=nnx.Rngs(0), precision="BF16_BF16_F32", dtype=jnp.bfloat16))

linear2 = jax.jit(
    nnx.Linear(
        5096, 5096, rngs=nnx.Rngs(0), precision="BF16_BF16_F32", dtype=jnp.bfloat16,
        dot_general=partial(jax.lax.dot_general, preferred_element_type=jnp.float32)
    )
)

linear3 = jax.jit(lambda x: linear(x).astype(jnp.float32))  # cast afterwards (may be optimized away by XLA)
linear4 = jax.jit(nnx.Linear(5096, 5096, rngs=nnx.Rngs(0)))  # default F32

x = jnp.ones((1024, 5096), dtype=jnp.float32)

# Precision
out = linear(x)    # jnp.bfloat16 (forced downcast)
out2 = linear2(x)  # jnp.float32
out3 = linear3(x)  # jnp.float32
out4 = linear4(x)  # jnp.float32

print(jnp.abs(out4 - out).mean())   # 0.00178 (worse precision)
print(jnp.abs(out4 - out2).mean())  # 0.00131 (better precision)
print(jnp.abs(out - out3).mean())   # 0.0

# Runtime (similar for BF16, much faster than F32)
# %timeit -> 369 μs ± 2.98 μs   (linear)
# %timeit -> 370 μs ± 0.64 μs   (linear2, with preferred_element_type)
# %timeit -> 388 μs ± 2.79 μs   (linear3)
# %timeit -> 683 μs ± 3.11 μs   (linear4, default F32)

### PR DESCRIPTION
**Pull request by vfdev-5**
# What does this PR do?

Fixes #4890

## Checklist
- [x] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [x] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: flax/linen/linear.py [hunk 1]
```diff
@@ -337,6 +337,7 @@ class Einsum(Module):
   kernel_init: Initializer = default_kernel_init
   bias_init: Initializer = initializers.zeros_init()
   promote_dtype: PromoteDtypeFn = promote_dtype
+  preferred_element_type: Dtype | None = None,
```
Comment (vfdev-5) [inline flax/linen/linear.py]: The change in linen is necessary for the tests nnx vs linen

File: flax/linen/linear.py [hunk 2]
```diff
@@ -337,6 +337,7 @@ class Einsum(Module):
   kernel_init: Initializer = default_kernel_init
   bias_init: Initializer = initializers.zeros_init()
   promote_dtype: PromoteDtypeFn = promote_dtype
+  preferred_element_type: Dtype | None = None
```
Comment (vfdev-5) [inline flax/linen/linear.py:340]: The change in linen is necessary for the tests nnx vs linen
